### DISCUSSION TITLE
Hypermodern Python podcast

### DISCUSSION THREAD

**Discussion by namurphy** [PlasmaPy/PlasmaPy#1533] [General]
I listened to a podcast recently on ["hypermodern" Python](https://talkpython.fm/episodes/show/362/hypermodern-python-projects) which was a really interesting discussion on a lot of the packaging & testing packages and infrastructure that we depend on.  There's also a [hypermodern Python package template](https://cookiecutter-hypermodern-python.readthedocs.io) which I played around with a little bit.  There are a bunch of things that we use already (like [pre-commit](https://pre-commit.com/), plus a few new things.

Some of the most interesting possibilities were...
 - Using [`poetry`](https://python-poetry.org/) for dependency management and packaging.
 - Automating uploads to [PyPI](https://pypi.org/) and [TestPyPI](https://test.pypi.org/)
 - Using [`nox`](https://nox.thea.codes/) instead of tox for testing environments.  The main difference is that the testing configuration is in a regular Python file for `nox` while it's in a `tox.ini` file for `tox`.
 - Static type checking with [mypy](http://mypy-lang.org/)
 - Runtime type-checking with [Typeguard](https://github.com/agronholm/typeguard)
 - Using a `src` layout for the package (i.e. move `plasmapy/` to `src/plasmapy`)
 - Checking documentation examples with [xdoctest](https://github.com/Erotemic/xdoctest)
 - Managing project labels with [GitHub Labeler](https://github.com/marketplace/actions/github-labeler)

Admittedly their list is pretty opinionated, but I was wondering if any of these would be worth trying out.  Some of the things might be relatively quick.  Switching from `tox` → `nox` might be ∼2 days of work, including updating the documentation and testing guide.  On the other hand, the type checking stuff would be really helpful in the long run but also a huge amount of work, since we'd need to change how to use our annotations for units and particles.

My reason for considering this now is that I've been dealing with a lot of packaging/testing/documentation infrastructure stuff for a couple of packages besides PlasmaPy, and in the long run it'd be easier to change things up sooner rather than later.  On the other hand, some of these changes might not be improvements given the tradeoffs, or worth the effort, or just low priority.  In any case, it is worth considering these different possibilities since some of them might really make code development easier in the long run.

**Comment by StanczakDominik** [comment]
> Automating uploads to [PyPI](https://pypi.org/) and [TestPyPI](https://test.pypi.org/)

We do have that! Only conditionally, though, on tagged commits: https://github.com/PlasmaPy/PlasmaPy/blob/fc08f0332530b8092f0a883eb0c696c7b0c69ac8/.github/workflows/testing.yml#L160-L175

> Static type checking with [mypy](http://mypy-lang.org/)

Would be nice to have be helpful, but is blocked on #1413. 

> Managing project labels with [GitHub Labeler](https://github.com/marketplace/actions/github-labeler)

That'd be neat to have. The manual web interface is meh at best.

The rest seem like they wouldn't help much...
